docs/wiki/concepts/Historical-Data.md

```diff
@@ -0,0 +1,146 @@
+- **`csp.item_at`**`(ts, index_or_time, duplicate_policy=DuplicatePolicy.LAST_VALUE, default=UNSET)`: returns tuple of `(datetime,value)` at requested `index_or_time`
+  - **`ts`**: the name of the input
+  - **`index_or_time`**:
+    - If providing an **index**, this represents how many ticks back to rereieve **and should be \<= 0**.
+      0 indicates the current value, -1 is the previous value, etc.
+    - If providing **time** one can either provide a datetime for absolute time, or a timedelta for how far back to access.
+      **NOTE** that timedelta must be negative to represent time in the past..
+  - **`duplicate_policy`**: when requesting history by datetime or timedelta, its possible that there could be multiple values that match the given time.
+    **`duplicate_policy`** can be provided to control the behavior of what to return in this case.
+    The default policy is to return the LAST_VALUE that exists at the given time.
+  - **`default`**: value to be returned if the requested time is out of the history bounds (if default is not provided and a request is out of bounds an exception will be raised).
+
+To illustrate the usage of history access using the **timedelta** indexing, consider a possible implementation of a function that sums up samples taken every second for each periods of **n_seconds** of the input time series.
+If the value ticks slower than every second then this implementation could sample the same value more than once (this is just an illustration, it's NOT recommended to use such implementation in real application as it could be implemented more efficiently):
+
+```python
+@csp.node
+def sample_sum(n_seconds: int, input: ts[int], default_sample_value: int = 0) -> ts[int]:
+    with csp.alarms():
+        a = csp.alarm(bool)
+    with csp.start():
+        assert n_seconds > 0
+        # This makes sure that input stores at least n_seconds seconds
+        csp.set_buffering_policy(input, tick_history=timedelta(seconds=n_seconds))
+        # Flag the input as passive since we don't need to react to its ticks
+        csp.make_passive(input)
+        # Schedule the first sample in n_seconds-1 from start, to also capture the initial value
+        csp.schedule_alarm(a, timedelta(seconds=n_seconds - 1), True)
+    if csp.ticked(a):
+        # Schedule the next sample in n_seconds from start
+        csp.schedule_alarm(a, timedelta(seconds=n_seconds), True)
+        res = 0
+        for i in range(n_seconds):
+            res += csp.value_at(input, timedelta(seconds=-i), default=default_sample_value)
+        return res
+```
+
+## Historical Range Access
+
+In similar fashion, the methods **`csp.values_at`**, **`csp.times_at`** and **`csp.items_at`** can be used to retrieve a range of historical input values as numpy arrays.
+The bin generator example above can be accomplished more efficiently with range access:
+
+```python
+@csp.node
+def data_bin_generator(bin_size: int, input: ts['T']) -> ts[['T']]:
```

Comment:
can we change this example to implement sample_sum using range access instead?
the example above is much less efficient, sum of the numpy buffer would be a better use here

Reply:
@pavithraes Ive gone ahead and updated the example for what I had in mind, I pushed to your branch

Reply:
Thank you!